src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs

```diff
@@ -0,0 +1,59 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using NuGet.Common;
+
+namespace NuGet.Protocol.Plugins
+{
+    public static class PluginDiscoveryUtility
+    {
+        public static string InternalPluginDiscoveryRoot { get; set; } = null;
+
+        public static string GetInternalPlugins()
+        {
+            var rootDirectory = InternalPluginDiscoveryRoot ?? System.Reflection.Assembly.GetEntryAssembly()?.Location;;
+
+            return rootDirectory ?? Path.GetDirectoryName(rootDirectory);
+        }
+
+        public static string GetNuGetHomePluginsPath()
+        {
+            var nuGetHome = NuGetEnvironment.GetFolderPath(NuGetFolderPath.NuGetHome);
+
+            return Path.Combine(nuGetHome,
+                "plugins",
+#if IS_DESKTOP
+                "netfx"
+#else
+                "netcore"
+#endif
+                );
+        }
+
+        public static IEnumerable<string> GetConventionBasedPlugins(IEnumerable<string> directories)
+        {
+            var paths = new List<string>();
+            foreach (var directory in directories.Where(Directory.Exists))
+            {
+                var pluginDirectories = Directory.GetDirectories(directory);
+
+                foreach (var pluginDirectory in pluginDirectories)
+                {
+#if IS_DESKTOP
+                    var expectedPluginName = Path.Combine(pluginDirectory, Path.GetFileName(pluginDirectory) + ".exe");
+#else
+                    var expectedPluginName = Path.Combine(pluginDirectory, Path.GetFileName(pluginDirectory) + ".dll");
+#endif
+
+                    var filesInDirectory = Directory.EnumerateFiles(pluginDirectory);
+                    if (filesInDirectory.Contains(expectedPluginName, PathUtility.GetStringComparerBasedOnOS()))
```

Comment:
can't you just use File.Exists since you already have the full path? Why do you need to enumerate files?

Reply:
File exists does not account for the casing. 
So it can get weird on Linux based FSs.

Another approach would be to just have an undefined behavior for the case sensitive use-cases. Which might be fine in our case anyways. 

I don't have a hard opinion one way or another, and the only thing required is for me to call out the behavior explicitly in the spec. 

What do you think? @rohit21agrawal

Reply:
Directoy.EnumerateFiles can take a major hit on perf, so i would rather ignore cases sensitive use-cases. In any case, the only part where casing could differ is "pluginName.exe/.dll" , which usually won't change casing given how we ship these assemblies.


If perf is not critical here, then i am fine with this.

Reply:
@jainaashish Unfortunately I wasn't able to find a better solution :)

I will rework it to where it doesn't consider casing at all, but it's more performant.